src/models/room.js

```diff
@@ -1277,19 +1291,37 @@ Room.prototype.updatePendingEvent = function(event, newStatus, newEventId) {
     } else if (newStatus == EventStatus.CANCELLED) {
         // remove it from the pending event list, or the timeline.
         if (this._pendingEventList) {
-            utils.removeElement(
-                this._pendingEventList,
-                function(ev) {
-                    return ev.getId() == oldEventId;
-                }, false,
-            );
+            const idx = this._pendingEventList.findIndex(ev => ev.getId() === oldEventId);
+            if (idx !== -1) {
+                const [removedEvent] = this._pendingEventList.splice(idx, 1);
+                if (removedEvent.getType() === "m.room.redaction") {
+                    this._revertRedactionLocalEcho(removedEvent);
+                }
+            }
         }
         this.removeEvent(oldEventId);
     }
 
     this.emit("Room.localEchoUpdated", event, this, oldEventId, oldStatus);
 };
 
+Room.prototype._revertRedactionLocalEcho = function(redactionEvent) {
+    const redactId = redactionEvent.event.redacts;
+    if (!redactId) {
+        return;
+    }
+    const redactedEvent = this.getUnfilteredTimelineSet()
+        .findEventById(redactId);
+    if (redactedEvent) {
+        redactedEvent.unmarkLocallyRedacted();
+        // re-render after undoing redaction
+        this.emit("Room.redaction", redactionEvent, this);
+        // reapply relation now redaction failed
+        if (redactedEvent.isRelation()) {
+            this._aggregateNonLiveEvent(redactedEvent);
```

Comment:
Hmm, does this re-add the relation to the relations collection, meaning there are now two copies of the relation, one that's still marked redacted and another live one? If so, I am worried that will get a bit confusing to keep track of.

Since we already have the relations collection listening for redaction, could it listen for unredaction and revive the event internally as opposed to duplicating it?

Reply:
At this point, the relation event shouldn't be aggregated in `Relations` anymore, as the local echo for the redaction emits `Event.beforeRedaction`, which causes `Relations::_onBeforeRedaction` to remove the relation. This is also why we need to add it again when the redaction is cancelled.

> Since we already have the relations collection listening for redaction, could it listen for unredaction and revive the event internally as opposed to duplicating it?

I considered this, but apart from this case, there is nothing interested in "unredaction" so it seemed a bit overkill to add an event for just this. The only duplication I can see here would be between aggregating a live event, and re-aggregating the relation on "unredaction" here. If we had an "unredaction" event, these paths wouldn't be unified, just these 3 lines would move to Relations... (+ the code for the unredaction event). This seems more direct to me, but the opposite would indeed be more symmetrical with the redaction event to me. Your choice :)

Reply:
Ah okay, you're right, it does get removed from the collection on `Event.beforeRedaction`, so I think it's fine as you have it.

I am starting to worry we've got little bits of redaction code spread all over the place, which feels hard to maintain, but that's not specific to this PR, so we can think about how to improve it separately.